src/ray/object_manager/object_manager.cc

```diff
@@ -145,38 +143,102 @@ ray::Status ObjectManager::Pull(const ObjectID &object_id) {
     RAY_LOG(ERROR) << object_id << " attempted to pull an object that's already local.";
     return ray::Status::OK();
   }
-  ray::Status status_code = object_directory_->SubscribeObjectLocations(
+  if (pull_requests_.find(object_id) != pull_requests_.end()) {
+    return ray::Status::OK();
+  }
+
+  pull_requests_.emplace(object_id, PullRequest());
+  return object_directory_->SubscribeObjectLocations(
       object_directory_pull_callback_id_, object_id,
       [this](const std::vector<ClientID> &client_ids, const ObjectID &object_id) {
-        RAY_CHECK_OK(object_directory_->UnsubscribeObjectLocations(
-            object_directory_pull_callback_id_, object_id));
-        GetLocationsSuccess(client_ids, object_id);
+        // Exit if the Pull request has already been fulfilled or canceled.
+        auto it = pull_requests_.find(object_id);
+        if (it == pull_requests_.end()) {
+          return;
+        }
+        // The timer was set if there were more clients to try. We record this
+        // now so that we can decide whether we should reset the timer when
+        // trying the new client locations.
+        bool timer_was_set = !it->second.client_locations.empty();
+        // Reset the list of clients that are now expected to have the object.
+        // NOTE(swang): Since we are overwriting the previous list of clients,
+        // we may end up sending a duplicate request to the same client as
+        // before.
+        it->second.client_locations = client_ids;
+        if (it->second.client_locations.empty()) {
+          // The object locations are now empty, so we should wait for the next
+          // notification about a new object location.
+          if (timer_was_set) {
+            // Cancel the timer if we were waiting to try the next client.
+            RAY_CHECK(it->second.retry_timer != nullptr);
+            it->second.retry_timer->cancel();
```

Comment:
If invoking `cancel` when a timer is not running does nothing, then consider invoking cancel whenever the timer has been instantiated and the client locations list is empty.

Reply:
Sounds good.